### PR TITLE
8284532: Memory leak in BitSet::BitMapFragmentTable in JFR leak profiler

### PR DESCRIPTION
**Pull request by zhengyu123**
BitSet::BitMapFragmentTable needs a destructor to cleanup its contents to avoid leaking hashtable entries.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284532](https://bugs.openjdk.java.net/browse/JDK-8284532): Memory leak in BitSet::BitMapFragmentTable in JFR leak profiler

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to 0ae03a26221ab385f72f54be20306c2b9db2e073
 * [Markus Grönlund](https://openjdk.java.net/census#mgronlun) (@mgronlun - **Reviewer**) ⚠️ Review applies to 0ae03a26221ab385f72f54be20306c2b9db2e073
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 0ae03a26221ab385f72f54be20306c2b9db2e073

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8148/head:pull/8148` \
`$ git checkout pull/8148`

Update a local copy of the PR: \
`$ git checkout pull/8148` \
`$ git pull https://git.openjdk.java.net/jdk pull/8148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8148`

View PR using the GUI difftool: \
`$ git pr show -t 8148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8148.diff">https://git.openjdk.java.net/jdk/pull/8148.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp [hunk 1]
```diff
@@ -29,6 +29,18 @@ BitSet::BitMapFragment::BitMapFragment(uintptr_t granule, BitMapFragment* next)
...
+BitSet::BitMapFragmentTable::~BitMapFragmentTable() {
+    for (int index = 0; index < table_size(); index ++) {
+      Entry* e = bucket(index);
+      while (e != nullptr) {
+        Entry* tmp = e;
+        e = e->next();
+        free_entry(tmp);
+      }
+    }
+    free_buckets();
```
Comment (tstuefe) [inline src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp:41]: Doesn't ~BasicHashtable do free_buckets already?

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi Zhengyu,

small remark inline, otherwise this looks okay. This should really be a utility function in BasicHashMap. 

What tests did you run?

Cheers, Thomas

Note that there is https://github.com/openjdk/jdk/pull/7964 open too, but it may be easier to get this one in first to keep it easy to backport.

File: src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp (hunk 1 above)
Comment (zhengyu123) [reply src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp]: Right. FIxed.

**Comment by zhengyu123** [comment]
> Hi Zhengyu,
> 
> small remark inline, otherwise this looks okay. This should really be a utility function in BasicHashMap.

Probably, I would like to defer it to separate CR, which it may touch quite a few places.

> 
> What tests did you run?
> 
Manual test with JFR recording `path-to-gc-roots=true`, then run `jcmd  JFR.dump`

> Cheers, Thomas
> 
> Note that there is #7964 open too, but it may be easier to get this one in first to keep it easy to backport.

**Comment by tstuefe** [comment]
> > Hi Zhengyu,
> > small remark inline, otherwise this looks okay. This should really be a utility function in BasicHashMap.
> 
> Probably, I would like to defer it to separate CR, which it may touch quite a few places.
> 

Sure. You just follow established patterns here.

> > What tests did you run?
> 
> Manual test with JFR recording `path-to-gc-roots=true`, then run `jcmd JFR.dump`

How did you notice the memory leak? In NMT dumps?

**Review by tstuefe — APPROVED** [review]
Looks good!

**Review by mgronlun — APPROVED** [review]
Looks good, thanks.

File: src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp [hunk 2]
```diff
@@ -29,6 +29,17 @@ BitSet::BitMapFragment::BitMapFragment(uintptr_t granule, BitMapFragment* next)
...
 
+BitSet::BitMapFragmentTable::~BitMapFragmentTable() {
+    for (int index = 0; index < table_size(); index ++) {
+      Entry* e = bucket(index);
+      while (e != nullptr) {
+        Entry* tmp = e;
+        e = e->next();
+        free_entry(tmp);
+      }
+    }
```
Comment (shipilev) [inline src/hotspot/share/jfr/leakprofiler/chains/bitset.cpp:40]: I think indenting is a bit off here: should be 2 spaces at all columns. Probably a copy-paste--from-destructor error.

**Review by shipilev — APPROVED** [review]
Looks fine, modulo a bit of indenting.

**Comment by zhengyu123** [comment]
Thanks, @tstuefe @mgronlun @shipilev

**Comment by zhengyu123** [comment]
> > > Hi Zhengyu,
> > > small remark inline, otherwise this looks okay. This should really be a utility function in BasicHashMap.
> > 
> > 
> > Probably, I would like to defer it to separate CR, which it may touch quite a few places.
> 
> Sure. You just follow established patterns here.
> 
> > > What tests did you run?
> > 
> > 
> > Manual test with JFR recording `path-to-gc-roots=true`, then run `jcmd JFR.dump`
> 
> How did you notice the memory leak? In NMT dumps?

Sorry, missed this comment earlier. No, I found it with `valgrind`

**Pull request closed without merging**